Port legacy all(*) to ATen

mrshenli opened the pull request:
Questions:

1. ~This PR disables `common_dtype` computation [in `TensorIterator.cpp`](https://github.com/mrshenli/pytorch/blob/all/aten/src/ATen/native/TensorIterator.cpp#L489-L491) for `all*` operators. The reason is that, [this code](https://github.com/mrshenli/pytorch/blob/all/aten/src/ATen/native/TensorIterator.cpp#L120) otherwise complains type mismatch, where the `op.tensor` is `type Variable[CPUByteType]` while the `op` is `CPUByteType`. I am not sure if this is the right solution for this problem.~

2. Should I clean up all occurrences of `_th_all` and `_th_all_out` (and `logicalAnd`, `logicalAndAll`)?

3. Do I need to implement derivatives for `all`?

@gchanan 

Benchmark:

<img width="590" alt="screen shot 2018-12-26 at 3 24 31 pm" src="https://user-images.githubusercontent.com/16999635/50456505-e9596a00-0922-11e9-844e-00c4b4aad7ca.png">

<img width="587" alt="screen shot 2018-12-26 at 3 26 10 pm" src="https://user-images.githubusercontent.com/16999635/50456509-ef4f4b00-0922-11e9-96bf-0a30c8574fe7.png">

<img width="590" alt="screen shot 2018-12-26 at 3 26 54 pm" src="https://user-images.githubusercontent.com/16999635/50456510-ef4f4b00-0922-11e9-8a63-e47988843cc8.png">

<img width="589" alt="screen shot 2018-12-26 at 3 27 16 pm" src="https://user-images.githubusercontent.com/16999635/50456511-ef4f4b00-0922-11e9-9004-2518aebcdc6e.png">

mrshenli committed:
port legacy all(*) to ATen

gchanan commented:
>> Should I clean up all occurrences of _th_all and _th_all_out (and logicalAnd, logicalAndAll)?

for the _th_ prefixed things, if you can.  For logicalAnd, logcialAndAll, it should be possible if you use `THTensor_wrap`.

>> Do I need to implement derivatives for all?

I don't think so, the return type should always have dtype uint8, which doesn't work with derivatives.

gchanan commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -947,21 +947,6 @@
-        - THTensor* self
-]]
 [[
   name: _th_all
```
can you get rid of this one?

mrshenli replied:
Yes, will do.

gchanan commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -126,7 +126,8 @@ struct CAFFE2_API TensorIterator {
 
   static std::unique_ptr<TensorIterator> binary_op(Tensor& out, const Tensor& a, const Tensor& b);
-  static std::unique_ptr<TensorIterator> reduce_op(Tensor& out, const Tensor& a);
+  static std::unique_ptr<TensorIterator> reduce_op(
```
can you avoid reformatting code that you don't touch?  It makes it harder to see what's going on for the reviewer.

mrshenli replied:
Yes, I will revert it.

gchanan commented on tools/autograd/derivatives.yaml:
```diff
@@ -138,6 +138,12 @@
 - name: alias(Tensor self)
   self: grad
 
+- name: all(Tensor self)
```
does this make any difference?

mrshenli replied:
I was hitting type mismatch (`Variable[CPUByteType]` vs CPUByteType) in `TensorIterator` without this change, because codegen is not unwrapping the input Tensor for `all(*)` without this line.

gchanan replied:
I see.  Can you write a comment here that says this is because TensorIterator doesn't work on Variables?  I'm unclear if that part is fixable, you can investigate in a future PR if you want.

gchanan commented on aten/src/ATen/native/cuda/ReduceOpsKernel.cu:
```diff
@@ -59,7 +59,7 @@ template <typename scalar_t, typename acc_t=scalar_t>
   gpu_reduce_kernel<scalar_t, scalar_t>(iter, func_wrapper<scalar_t> ([]GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
     return a * b;
-  }), 1);
+  }), true);
```
did you mix this up?  Why is the identity of prod `true` but the identity of and `1`?

mrshenli replied:
ah! my bad...

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -448,6 +452,30 @@ Tensor norm(const Tensor& self, Scalar p) {
+    "all only supports torch.uint8 dtype");
+
+  Tensor result = at::empty({0}, self.options());
+  ScalarType dtype = get_dtype(result, self, {}, true);
```
why call get_dtype?  You always know the answer is going to be result.dtype.  The arguments to this also don't look correct; you don't want to promote_integers for all -- there's no risk of overflow.

gchanan commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -461,7 +489,9 @@ Tensor &all_out(Tensor &result, const Tensor &self, int64_t dim, bool keepdim) {
     return result;
   } else {
-    return at::legacy::th::_th_all_out(result, self, dim, keepdim);
+    ScalarType dtype = get_dtype(result, self, {}, true);
```
`get_dtype` could make sense here but I don't think you want promote_integers to be true.

mrshenli commented:
Operator `all(*)` on gpu and  `all(1, keepdim=1)` on cpu is significantly faster than the original code, while `all(*)` on cpu is ~30% slower, which makes me doubt the correctness of this PR. Let me dig and add some tests.

mrshenli committed:
Address comments

mrshenli commented on aten/src/ATen/native/cuda/ReduceOpsKernel.cu:
```diff
@@ -119,9 +126,14 @@ static void mean_kernel_cuda(TensorIterator& iter) {
 }
 
+static void and_kernel_cuda(TensorIterator& iter) {
+  and_kernel_impl(iter);
```
Have to do the wrapping here, otherwise it hits the following error on windows:

```
error : On Windows, the enclosing parent function ("and_kernel_cuda") for an extended __host__ __device__ lambda cannot have internal or no linkage [C:\Jenkins\workspace\caffe2-builds\py2-cuda9.0-cudnn7-windows-build\build\caffe2\caffe2_gpu.vcxproj]
```

mrshenli committed:
clean up _th_all

mrshenli committed:
add tests

mrshenli commented:
I added tests for `cpu` and `cuda`, and the results seem to be correct. 

@gchanan 
Is it normal to see such substantial performance improvement by just porting code?

gchanan commented:
can you redo the benchmarks with timeit? One you can see a simple example I did here: https://github.com/pytorch/pytorch/pull/15484

If the CPU code wasn't parallelized before you should see a big difference, but you would also want to compare to the non-parallelized code (via setting OMP_NUM_THREADS=1).

mrshenli commented:
Sure, will do!

mrshenli commented:
@gchanan 

The results confirm with the previous benchmark. I will do this for other PRs as well.

now, with `OMP_NUM_THREADS` unset

```ipython
In [7]: x = torch.rand(100, 100, 100) > 0.5

In [8]: %timeit x.all()
111 µs ± 2.86 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [9]: %timeit x.all(1, keepdim=True)
118 µs ± 12.6 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [10]: %timeit x.all(2)
99.7 µs ± 1.07 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [2]: x = torch.rand([100, 100, 100], device='cuda') > 0.5

In [3]: %timeit x.all()
24.9 µs ± 608 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [4]: %timeit x.all(1, keepdim=True)
17.7 µs ± 16.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit x.all(2)
32.4 µs ± 5.65 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

now, with `export OMP_NUM_THREADS=1`

```
In [2]: x = torch.rand(100, 100, 100) > 0.5

In [3]: %timeit x.all()
1.09 ms ± 9.08 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [4]: %timeit x.all(1, keepdim=True)
1.13 ms ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [5]: %timeit x.all(2)
1.13 ms ± 12.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

before

```
In [3]: x = torch.rand(100, 100, 100) > 0.5

In [4]: %timeit x.all()
63.1 µs ± 101 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]: %timeit x.all(1, keepdim=True)
16.8 ms ± 161 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: %timeit x.all(2)
16.7 ms ± 121 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [7]: x = torch.rand([100, 100, 100], device='cuda') > 0.5

In [8]: %timeit x.all()
106 µs ± 5.04 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [9]: %timeit x.all(1, keepdim=True)
16.4 µs ± 1.6 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [10]: %timeit x.all(2)
51.2 µs ± 247 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

fmassa commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -307,6 +307,15 @@ template <class T> Vec256<T> inline operator/(const Vec256<T> &a, const Vec256<T
+    const Vec256<T> &a, const Vec256<T> &b) {
+  Vec256<T> c = Vec256<T>();
+  for (int i = 0; i != Vec256<T>::size(); i++) {
+    c[i] = a[i] && b[i];
```
Question: are there vectorized bitwise and operators that you could use for avx2 etc?

mrshenli replied:
Good point! Let me check.

mrshenli replied:
@fmassa  @gchanan 
Let me try something as follows (will make generic and use loops):

```c++
Vec256<uint8_t> inline operator&&(const Vec256<uint8_t> &a, const Vec256<uint8_t> &b) {
  a = _mm256_or_si256(a, _mm256_srl_epi64(a, 4));
  a = _mm256_or_si256(a, _mm256_srl_epi64(a, 2));
  a = _mm256_or_si256(a, _mm256_srl_epi64(a, 1));
 
  // do the same for b

  return  mm256_and_si256(0x01010101..., mm256_and_si256(a, b));
}
```

Will rerun benchmark after this change. Comments are appreciated!

mrshenli committed:
implement vec256<uint8_t> operator&& with avx2

mrshenli commented on aten/src/ATen/cpu/vec256/vec256_bool.h:
```diff
@@ -0,0 +1,43 @@
+};
+
+template <>
+Vec256<uint8_t> inline operator&&(
```
Pushing this up for testing. I am trying to figure out how to trigger this in my local test. 

This PR is only doing this for `uint8_t` for now. If this works and does speed it up, I will add another PR to implement similar algorithm for other `vec256_*.h` files.

`Vec256<uint8_t>` is not directly using `__m256i ` as its value, as I am worried that its `blend` and `set` methods might need to use the sign bit of `int64_t`, which I think is not wrong but is a little weird. 

@fmassa @gchanan any suggestion?

mrshenli commented on aten/src/ATen/cpu/vec256/vec256_bool.h:
```diff
@@ -0,0 +1,43 @@
+  mb = _mm256_or_si256(mb, _mm256_srli_si256(mb, 2));
+  mb = _mm256_or_si256(mb, _mm256_srli_si256(mb, 1));
+
+  _mm256_storeu_si256(
```
Don't need a mask here, as the [doc](https://software.intel.com/en-us/node/523973) says the high-order bits will be cleared to 0.  (BTW, I think there are some typos in the doc. e.g., low-order -> high-order)

Will add this to comments, in final revision.

mrshenli commented:
The performance does improve after switching to AVX2. Thanks @fmassa !

```python
In [2]: In [7]: x = torch.rand(100, 100, 100) > 0.5

In [3]:  %timeit x.all()
44.9 µs ± 5.2 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [4]: %timeit x.all(1, keepdim=True)
92.9 µs ± 4.52 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [5]:  %timeit x.all(2)
84.9 µs ± 6.29 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

mrshenli committed:
Address comments:

1. remove logicalAndAll and logicalAnd
2. move tests to test_torch.py

gchanan commented on aten/src/ATen/cpu/vec256/vec256_bool.h:
```diff
@@ -0,0 +1,331 @@
+    return _mm256_and_si256(one, _mm256_cmpeq_epi8(min, values));
+  }
+  Vec256<uint8_t> operator>(const Vec256<uint8_t>& other) const {
+    // cannot use _mm256_cmpgt_epi8 which is for signed bytes
```
do we need this now?

mrshenli replied:
I think yes. 

I think it is more natural/readable to use `_mm256_cmpgt_epi8` instead of `max`/`eq` operations. But, sadly AVX2 does not have a unsigned version for `>`.

gchanan commented on aten/src/ATen/cpu/vec256/vec256_bool.h:
```diff
@@ -0,0 +1,331 @@
+#pragma once
```
can you pull out the vec256_bool stuff into another PR?

uint8 is strange because sometimes we treat it as a an actual uint8 and sometimes we treat it as a bool, and that's something we should fix in the future.  From a quick read of this I can't tell which parts implement the uint8_t behavior and which parts implement the bool behavior.  Is there bool-specific behavior here?

mrshenli replied:
Sure, let me pull it out.

mrshenli committed:
Address Comments:

1. pull out vec256_bool implementation
2. add comments to derivatives.yaml to explain the reason for adding all(*)

mrshenli committed:
edit comments

colesbury commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -307,6 +308,15 @@ template <class T> Vec256<T> inline operator/(const Vec256<T> &a, const Vec256<T
+    const Vec256<T> &a, const Vec256<T> &b) {
+  Vec256<T> c = Vec256<T>();
+  for (int i = 0; i != Vec256<T>::size(); i++) {
+    c[i] = a[i] && b[i];
```
I don't think it's a good idea to add this to Vec256:

1) The behavior differs from the other "boolean" type operations (such as comparisons). In those, the result for true is all ones (which is -1 for integers and a NAN for floats). Here the true result is part zeros part ones.

2) There's only a non vectorized version, which means that it's better to not use Vec256 here at all.

I realize it's difficult to avoid the Vec256 because binary_kernel_reduce_vec expects it. I see two possible options:

a) Add a `binary_kernel_reduce` that only operates on scalars. You can use `binary_kernel_reduce_vec` as a starting point

b) Add this only to and_kernel_impl instead of Vec256. Even if it's slower than it needs to be, at least it doesn't add an operator that has inconsistent behavior with the other operations.

mrshenli replied:
Thanks, will address.

Question:

Regarding the inconsistency, I think the code might still be correct if we change `&&` to return `-1` as other logical operations, as eventually it reduces to a single scalar, which will use the scalar version `&&` (haven't tried). The downside is that is it using an assumption on the internal implementation of `binary_kernel_reduce` which is not explicitly advertised.  Is this acceptable?

mrshenli committed:
remove operator&& from vec256 to avoid inconsistency with other logical operators

mrshenli committed:
Merge remote-tracking branch 'upstream/master' into all

mrshenli committed:
Merge remote-tracking branch 'upstream/master' into all

mrshenli committed:
fix typo

mrshenli commented:
The failed test passes after [rerun](https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-test/7562/).

gchanan commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -54,11 +54,29 @@ static void prod_kernel_impl(TensorIterator& iter) {
+    iter,
+    [=](uint8_t a, uint8_t b) -> uint8_t { return a && b; },
+    [=](Vec256<uint8_t> a, Vec256<uint8_t> b) {
+      // adding the implementation here instead of in vec256_base because it is
```
nit: I think this comment could go into a little more depth; it's not immediately obvious why the -1/0 distinction matters.

mrshenli replied:
Will edit before land. Thanks!

mrshenli committed:
address comments

mrshenli committed:
Merge remote-tracking branch 'upstream/master' into all
